Navigation Link: Use block bindings to dynamically fetch the URL in the editor and on front end.

getdave opened the pull request:
## What

This PR improves the Navigation Link block in two key ways:

* makes Navigation Link blocks' URL attribute _dynamic_ if the link is to an entity (e.g. Page, Post...etc) using the Block Bindings API
* makes it clear when a Nav Link block is linking to an _entity_ vs a static URL through improved UI indicators

Closes #18345

## Why

### The Core Problem

The Navigation Link block has never dynamically fetched URLs for linked entities (e.g. Pages, Posts):

* When a user selects a Page/Post in the Editor, the block stores:
  * `id`: the ID of the entity.
  * `url`: a static copy of the permalink at that time.
* This url remains fixed — even if the entity's actual permalink changes later.
* Critically, the front end also uses this static URL. There is no dynamic lookup on render, which was previously assumed.

This means:

* If you change a page's URL (e.g. change the slug), existing Nav Links break silently.
* There is a false sense of safety — users believe they are linking to an entity, but the system never tracks or updates the link.

### Why This Is a Major Usability Issue

1. **WYSIWYG Mismatch**
   * Editor shows outdated links.
   * Front end shows exactly the same outdated links.
   * This contradicts user expectations and introduces subtle bugs.

2. **Invisible Link Types**
   * There's no visual difference between a link to a Page and a manually entered static URL.
   * Users can't tell if their link is "safe" (entity-bound) or "fragile" (static).

3. **Poor Content Resilience**
   * If you rename or move content, your menus don't adapt.
   * This leads to 404s and broken navigation — especially problematic on large or client-managed sites.

## How

Using the Block Bindings API, we bind the `url` attribute of the block to the entity represented by the `id` attribute of the block through a new `core/entity` binding source.

Key implementation details:

* Created a generic `core/entity` binding source that uses `source_args['key']` to determine which property to return (e.g., `url`)
* Updated the Navigation Link block to automatically create bindings when an entity is selected
* Enhanced the `LinkControl` UI to clearly indicate when a link is bound to an entity
* Disabled manual URL editing for bound links with an "unlink" option to break the binding
* Updated PHP rendering to use the Block Bindings API for automatic URL resolution

This creates a unified approach between editor and frontend, with clear visual indicators for dynamic vs static links.

**Note:** After discussion with subject matter experts (@ockham and @cbravobernal), we agreed to implement a generic `core/entity` binding source for now. @ockham will investigate the possibility of normalizing this with existing data sources in the future.

## Proposed Followups

The following items are being pushed to immediate follow ups:

* update design to follow mockups in [this comment](https://github.com/WordPress/gutenberg/pull/71630#issuecomment-3337893348)
* automatically update the `Text` content of the Nav Link item to match the chosen entity. This is the most likely workflow.
* update the Submenu block to match the implementation. Requires a quick refactor to align and share code between the blocks first to make this refactor easier.
* remove `entity` block binding and utilise existing post-data and term-data bindings once they have been made more extensible (see [this comment](https://github.com/WordPress/gutenberg/pull/71630#issuecomment-3354889752))
* [Avoid using state to manage focus](https://github.com/WordPress/gutenberg/pull/71630#discussion_r2397279884)

getdave committed:
Initial commit

talldan commented on packages/block-editor/src/utils/block-bindings.js:
```diff
@@ -16,6 +16,7 @@ const BLOCK_BINDINGS_ALLOWED_BLOCKS = {
 	'core/heading': [ 'content' ],
 	'core/image': [ 'id', 'url', 'title', 'alt', 'caption' ],
 	'core/button': [ 'url', 'text', 'linkTarget', 'rel' ],
+	'core/navigation-link': [ 'url' ],
```
I guess `navigation-submenu` will need this as well.

getdave replied:
It absolutely will 👍  It's one of many things that either are or should be shared between the blocks. I intend to create a `/shared` directory in the Nav Link block and place all shared stuff there to make it ultra clear and avoid any potential for confusion on impact.

cbravobernal replied:
@ockham is working on a single source of truth for supported block attributes.
https://github.com/WordPress/gutenberg/pull/71820

We plan to land it asap, so this may be need to be updated later.

cbravobernal replied:
Did you skip the PHP registration?
Something like:
https://github.com/WordPress/gutenberg/blob/76105c84fc85b61335dbefc38d340267366ba165/lib/compat/wordpress-6.9/block-bindings.php#L11-L32

getdave replied:
It's now ✅

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -231,8 +231,30 @@ function render_block_core_navigation_link( $attributes, $content, $block ) {
+	$url = $attributes['url'] ?? '';
+
+	// Handle block bindings for URL - check if there's a binding and resolve it
+	if ( isset( $attributes['metadata']['bindings']['url'] ) ) {
```
I don't think you should need to do this, block bindings should automatically inject the right attribute value into the render callback. 🤔

getdave replied:
Yeh I was wondering about that. Let's try it out without it and see what happens.

getdave replied:
> I don't think you should need to do this

...thinking again. By "this" did you mean the whole process of resolving the value? Would this happen automatically and all we need to do is render the `url` attribute?

cbravobernal replied:
`process_block_bindings` should do the trick. I tested and seems to work on my end.

Text file: [bb_patch.txt](https://github.com/user-attachments/files/22563087/bb_patch.txt)

Patch file: 
[bb_patch.patch](https://github.com/user-attachments/files/22563158/bb_patch.patch)

getdave replied:
In my testing this seemed to work without having to manually resolve the binding 🤔 I simplified the implementation accordingly. 

When this PR is reviewed and tested we'll see if this has broken anything.

talldan commented on packages/editor/src/bindings/api.js:
```diff
@@ -26,4 +27,5 @@ export function registerCoreBlockBindingsSources() {
 	registerBlockBindingsSource( postData );
 	registerBlockBindingsSource( postMeta );
 	registerBlockBindingsSource( termData );
+	registerBlockBindingsSource( entityUrl );
```
Out of interest, do the existing `postData` / `termsData` bindings work for your use case?

I guess your issues are probably:
a) they are read/write
b) only supports terms or only supports posts
c) I'm not sure if post data server-side includes a 'link' property, `get_permalink` might be a requirement.

It'd be interesting to know if these things are solveable. To me, it feels like they might be.
a) possibly you could `return false` from `canUserEditValue` when the property is `link`.
b) you could bind only to `post-data` when it's a post type being linked, and only to `term-data` when it's a term.
c) maybe the server side implementation of the bindings could have some special case for `link` if this is an issue.

edit: just to mention I don't know much about these binding sources, they're new since I last looked at it, so I might be wrong.

getdave replied:
I thought the same. I tried using `postData` but it uses the current post ID from context so that didn't fit the use case.

Our version uses the block's own `attributes` to create the binding. I think we may want to make this a generic binding as in future I see this also being required for `core/paragraph` and other places where the link format (i.e. the one from `@wordpress/formats`) is used.

cbravobernal replied:
Thought the same here. But I guess I have a lack of context. We should avoid creating custom sources for every use case though 😅

getdave replied:
Just to update please see https://github.com/WordPress/gutenberg/pull/71630#issuecomment-3352284109

getdave commented:
## Status Update

The more I see this the more I'm convinced we need to decouple the work on `LinkControl` and pull it out into it's own PR. Essentially we need LinkControl to 

- have an `Boolean` to turn entity handling on/off
- have it use a different UI internally based on whether the supplied `value` has an `id` 

The specifics of the above are TBC but they are prerequisites for landing the binding on the Nav block and it's making this PR quite complex and large.

getdave commented on packages/block-library/src/navigation-link/block.json:
```diff
@@ -44,6 +44,9 @@
 		"isTopLevelLink": {
 			"type": "boolean"
+		},
+		"metadata": {
+			"type": "object"
 		}
```
@talldan I wonder if we need this? Is it auto-added when you initialise a block binding? 🤔

cbravobernal replied:
Bindings have this structure:

```html
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/entity-url"}}}} -->
<p></p>
<!-- /wp:paragraph -->
```

So is quite probable that, on the other hand, neither the image or the paragraph contains that block json metadata object defined, so is probably wrong.

Neither this example PR
https://github.com/WordPress/gutenberg/pull/70975

getdave replied:
Thanks. They are not required.

getdave commented on packages/editor/src/bindings/entity.js:
```diff
@@ -0,0 +1,60 @@
+			);
+			url = term?.link || '';
+		}
+
```
We probably need to magically handle `post-type-archive` links as well if that's something we even want to do.

For example, if you link to the `Books` Post archive and then you rename the CPT to `Paper` should you expect the link to automatically update or would that be taking things too far?

scruffian commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -406,8 +418,23 @@ function LinkControl( {
+								if ( isEntity ) {
+									return (
+										<Button
+											variant="tertiary"
```
```suggestion
```
This makes it blue. I think its better to use just a Button with no variant:

<img width="345" height="341" alt="Image" src="https://github.com/user-attachments/assets/05fa0cf8-5380-41b2-91ac-07e6ff310509" />

scruffian commented:
I think there's a focus loss when clicking the "unlink" button too:

https://github.com/user-attachments/assets/3415cb3b-8bb4-4947-aaa8-ffe034280cf3

I think when you click "unlink" the cursor should land back in the URL field. It would be good to get a sanity check from @jeryj though.

scruffian commented:
When the submenu blocks are in a synced state, they don't get the purple outline, or icon:
<img width="671" height="487" alt="Screenshot 2025-09-26 at 11 48 22" src="https://github.com/user-attachments/assets/04708a58-7214-4624-8777-513aeb169bf0" />

I think this is also true for links that are synced but haven't been set through the new UI

cbravobernal commented on lib/compat/wordpress-6.9/entity-url-block-bindings.php:
```diff
@@ -0,0 +1,77 @@
+		return null;
+	}
+
+	$entity_id = $source_args['id'];
+	$type      = $source_args['type'] ?? '';
+	$kind      = $source_args['kind'] ?? '';
```
Can you use `uses_context` in this case?

Something like:
https://github.com/WordPress/gutenberg/blob/3e2da58d8940eb7e738746a6e03040393c57bf0b/lib/compat/wordpress-6.9/term-data-block-bindings.php#L91

That way you may not need to add extra source arguments.

cbravobernal replied:
I've seen some duplication between the attributes and the source args. You have access to those attributes in the callback if needed.

```suggestion
  $entity_id = $block_instance->attributes['id'];
	$type      = $block_instance->attributes['type'] ?? '';
	$kind      = $block_instance->attributes['kind'] ?? '';
```

getdave replied:
I think we should use the source args. This avoids use coupling this binding specifically to the contract of the Nav Link block's attribute definition. Do you agree?

cbravobernal replied:
I see some duplication in the final code in the HTML 😅 

```
<!-- wp:navigation-link {"label":"Custom page.","type":"page","id":400,"url":"http://localhost:8888/lorem-ipsum/","kind":"post-type","metadata":{"bindings":{"url":{"source":"core/entity-url","args":{"id":400,"type":"page","kind":"post-type"}}}}} /-->
```

I may lack of context, but I still have some doubts about it.

getdave replied:
We have two options as I see it

1. Use the block attributes to get the `id` which represents the entity. We can then ditch al the bindingArgs thus removing redundancy in the final block markup but reducing flexibility of the binding itself.
2. Use the bindArgs as we have it which makes the binding more flexible for other (non-Nav block) contexts but increases markup redundancy.

Perhaps we need to accept that we should make the binding block specific (at least for now) and use the block's `attributes`? I'd like to arrive at a consensus  so appreciate your input 🙇

cbravobernal replied:
> Use the bindArgs as we have it which makes the binding more flexible for other (non-Nav block) contexts...

Can you please provide an example of this use case? I cannot imagine one right now 😓

scruffian replied:
> Can you please provide an example of this use case? I cannot imagine one right now 😓

We have talked about doing similar URL bindings for the buttons block and any links.

scruffian commented:
> I have pushed what I hope is a fix for this. Could you pull and re-test 🙏

I can confirm this is fixed.

> I think I came across this too, but I can no longer replicate it. Can you?

I can't replicate this anymore.

cbravobernal commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -393,6 +477,33 @@ export default function NavigationLinkEdit( {
+					! blockBindingsSource?.canUserEditValue?.( {
+						select,
+						context,
+						args: metadata?.bindings?.url?.args,
```
We may not need args.

cbravobernal commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -703,6 +822,23 @@ export default function NavigationLinkEdit( {
+										url: {
+											source: 'core/entity-url',
+											args: {
+												id: updatedValue.id,
+												type: updatedValue.type,
+												kind: updatedValue.kind,
+											},
+										},
+									} );
+								}
```
```suggestion
```
We may not need source args for this case if we just read the attribute id.

getdave replied:
I've been thinking...

Using block attrs makes this binding very use case specific and not reusable. 

If we were to use binding args then we could make this generic and allow look up of any entity based on the binding args. 

WDYT?

getdave replied:
FYI I moved to bindingArgs but now I'm not convinced it's the best approach. I tried to [summarise things in this comment](https://github.com/WordPress/gutenberg/pull/71630#discussion_r2387082588).

scruffian replied:
I believe in the end we decided to the block attributes and go with the specific solution, at least in the first case. We can reconsider later if we want to make this apply to other blocks.

cbravobernal reviewed (commented):
Thanks for your PR. Is really exciting to see how Block Bindings can fix some bugs in Core, and it is awesome that you discovered they could be a solution 🚀 

PD: Ping @ockham and @cbravobernal, review the backport to make sure we land the new supports correctly.

getdave commented on packages/editor/src/bindings/api.js:
```diff
@@ -10,6 +10,7 @@ import patternOverrides from './pattern-overrides';
 import postData from './post-data';
 import postMeta from './post-meta';
 import termData from './term-data';
+import entityUrl from './entity-url';
```
This binding is for `entityUrl` but int he future I can imagine wanting other fields from the entity. For example `title` and binding them to other block attributes (e.g. `label`). Could we have an single binding that returns the _entire_ entity object?

getdave commented:
> Thanks for your PR. Is really exciting to see how Block Bindings can fix some bugs in Core, and it is awesome that you discovered they could be a solution 🚀


I spoke directly with @ockham and @cbravobernal and we landed on the following:

- storing "data" in `source_args` isn't optimal and increases duplication
- we should revert back to using `block.attributes`
- in the future we can try a fallback approach whereby we look for `source_args` and fallback to block `attributes` if source args aren't defined. This will allow the binding to be used in other contexts (e.g. button block...etc)
- we can’t easily reuse the existing post-data binding that @ockham has worked on due to its lack of taxonomy support and reliance on block context - stretch goal for 6.9 would be to resolve this problem and have a single unified implementation.
- for now refactor our binding to a core/entity block binding which uses sourceArgs['key'] to determine which property we need from the entity (e.g. url).

getdave commented on packages/editor/src/bindings/entity.js:
```diff
@@ -0,0 +1,74 @@
+		}
+		// Handle taxonomies
+		else if ( kind === 'taxonomy' ) {
+			// Convert 'tag' back to 'post_tag' for API calls
+			// See update-attributes.js line 166 for the reverse conversion
+			const taxonomySlug = type === 'tag' ? 'post_tag' : type;
```
This is very awkward and creates problems. We need to be able to trust the `type` as matching whatever we need to use to query the REST API.

getdave replied:
See https://github.com/WordPress/gutenberg/issues/71979

getdave commented:
> > we can’t easily reuse the existing post-data binding that @ockham has worked on due to its lack of taxonomy support
> 
> 
> 
> There is a `core/terms-data` binding source, could that provide what's needed for taxonomy support? The block could switch between sources depending on what the user links.
> 
> 
> 
> There's still the context issue. Sounds like it's best tackled in a separate PR to explore that. 👍 

Yes agreed. Having discussed with @ockham  and @cbravobernal (see comment) we also noticed the potential overlap but they were (as yet) unable to suggest as way around the context problem. So we've agreed to run with entity and then circle back to normalizing towards existing sources. 

As you say this can be a follow up which should ideally land in 6.9. 

Did you get a chance to ask this PR for a spin? More eyes on the functionality to catch any regressions or bugs would be very much appreciated.

getdave commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -231,8 +231,11 @@ function render_block_core_navigation_link( $attributes, $content, $block ) {
 	// Start appending HTML attributes to anchor tag.
-	if ( isset( $attributes['url'] ) ) {
-		$html .= ' href="' . esc_url( block_core_navigation_link_maybe_urldecode( $attributes['url'] ) ) . '"';
+	// Note: The URL may be bound via block bindings and automatically resolved
+	$url = $attributes['url'] ?? '';
+
+	if ( ! empty( $url ) ) {
+		$html .= ' href="' . esc_url( block_core_navigation_link_maybe_urldecode( $url ) ) . '"';
```
This seems a little pointless as a refactor. Let's remove it.

getdave commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -31,6 +31,7 @@ Right now, not all block attributes are compatible with block bindings. There is
 | Heading             | content                    |
 | Image               | id, url, title, alt        |
 | Button              | text, url, linkTarget, rel |
+| Navigation Link     | url                        |
```
https://github.com/WordPress/gutenberg/pull/71994 is going to allow us to use the same functionality on Submenu block. We need to land that one first

getdave replied:
Also 

https://github.com/WordPress/gutenberg/pull/72002
https://github.com/WordPress/gutenberg/pull/72004

talldan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -336,6 +342,29 @@ function LinkControl( {
+		setShouldFocusSearchInput( true );
+	};
+
+	// Focus the search input when requested, once the component has re-rendered
+	// This ensures the input is enabled and ready to receive focus
```
Is this the only way? What about `setTimeout` or `requestAnimationFrame` from the event handler?

This feels like a lot because it causes a component re-render just to do a DOM manipulation.

getdave replied:
Yes it's really not ideal. Given you've ✅  I'm hoping resolving this can be one of the follow ups. I'll add to the list in the PR desc 🙇

talldan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -420,8 +464,8 @@ function LinkControl( {
-								)
-							}
+								);
+							} )() }
```
Maybe it could be a component if it has reached a certain level of complexity, an IIFE feels like an unexpected pattern to see here, I can't recall seeing it before in the codebase.

getdave replied:
Yes it's got out of the control! I'll refactor that now.

getdave replied:
I extracted a simple component for now. I don't like the component and it needs refactoring later to make it better.

talldan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -142,6 +143,7 @@ function LinkControl( {
 	hasRichPreviews = false,
 	hasTextControl = false,
 	renderControlBottom = null,
+	handleEntities = false,
```
Very minor, but I wonder if the terminology should maybe be `binding` rather than `entities`, as I see `hasUrlBinding` being used elsewhere in the PR.

getdave replied:
I've had a think about this. For me the concept of a "binding" is alien to the LinkControl. It doesn't care about bindings. All it cares about is whether the `value` has an `id` and thus represents an entity (Page, Post, Taxonony...etc).

I'm hoping this isn't a bad design decision, but for me the concept LinkControl needs to care about is "am I representing an entity or a static URL". If it's the former (denoted by presence of `id` in `value`) then use a different UI/UX to make that clear.

I don't want it to have to understand about a whole new thinking called "bindings".

Btw this flag exists as an opt in so that other instances of `LinkControl` don't suddenly start trying to use this new UX when we haven't prepped for it.

talldan commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -121,6 +122,15 @@ const LinkControlSearchInput = forwardRef(
+		const helpText = isEntity
+			? sprintf(
+					/* translators: %s: entity type (e.g., page, post) */
+					__( 'Link stays in sync with the selected %s.' ),
```
Part of me wonders if it might be better not to show the URL at all when a binding is active. Or to show the Post/Page name. 🤔 

I think the way it is will be good for the first iteration, though, and maybe it's weird without the URL. The text could be shortened to 'Synced with the selected $s.' (or again use the post/page title), but these are really minor ponderings as the whole thing works very well in practice.

getdave replied:
Yes. I agree. 

Design input was the fix is critical enough that we should ship "as is" and refine afterwards.

@jasmussen has provided [a mockup which aligns with your suggestion](https://github.com/WordPress/gutenberg/pull/71630#issuecomment-3337893348). 

I've noted several proposed followups in the PR description so we don't forget them.

talldan commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -393,6 +477,11 @@ export default function NavigationLinkEdit( {
+	// URL binding logic
+	const { updateBlockBindings } = useBlockBindingsUtils( clientId );
+
+	const hasUrlBinding = isSelected && !! metadata?.bindings?.url && !! id;
```
I'm curious why the block only `hasUrlBinding` when `isSelected`?

getdave replied:
I must have introduced this accidentally. I've been doing a lot of rebases to keep this PR up to date with changes I made in `trunk`.

Anyway, good spot and I removed it! 😅

talldan reviewed (approved):
This is looking good to me and tests really well, I only had a few minor non-blocking comments. It might be nice if someone who's worked more on block bindings recently give it an approving review too. Up to you as the author though.

It also makes me wonder if we should offer to bind the post/page title for the user as another feature.

E2E tests would be good.

getdave committed:
Post rebase fixes

getdave committed:
Improve help text as per https://github.com/WordPress/gutenberg/pull/71630\#discussion_r2397328702

getdave committed:
Extract simple suffic component

getdave committed:
Remove erroneous isSelected condition for url binding condition

getdave committed:
Initial e2e test coverage

getdave commented on packages/block-library/src/navigation-link/shared/controls.js:
```diff
@@ -104,6 +160,30 @@ export function Controls( {
+							);
+						} )()
+					}
+					suffix={
+						hasUrlBinding && (
+							<Button
+								icon={ unlinkIcon }
+								onClick={ editBoundLink }
+								aria-label={ __( 'Unlink and edit' ) }
+								__next40pxDefaultSize
+							/>
```
@jeryj This button allows users to remove the binding and unlock the input to change the URL. How can we properly communicate this relationship in an accessible manner?

Same question for the identical interface within the `LinkControl`.

jeryj replied:
I'm going to test it out for a bit to get a better understanding of the full context. One thing that jumps out is the use of "unlink" when we're editing links. Maybe "unsync" would be better? Would visible help text be better for everyone to understand this action?

jeryj replied:
Tested it out proper. Here's my thoughts:
1. "Unlink" should be "Unsync": Unlink in this context means more "remove the link" rather than Unsync the link.
2. The "Synced with the selected page" help text underneath this input's aria-descridby isn't attached to the button, it's attached to an unfocusable input, so screen reader users will likely skip it. If we're in this synced state, the aria-described by text should get assigned to the button so it's accessible. 
3. This button (what I think should be "Unsync") should have a tooltip that says "Unlink and edit." I don't think the action here is clear enough.
4. Rather than say, "Synced with the selected %s.' it would be far more helpful to say something like "Synced with <a href="pagelink">pagename</a>." Hopefully translated link interpolation here isn't too much of a pain.

Also, potentially not related to this PR as we've had it happen before - keyboard navigation is broken on the sidebar inspector for the navigation block. If you press Tab from the Text input, it moves focus to the canvas instead of the next item in the sidebar.

https://github.com/user-attachments/assets/3ed8ea2f-96d5-4bb5-b938-bf5bb52ab5ec

getdave replied:
That all sounds doable. Thanks for the insight. This is why e2e tests are useful as they exposed this issue to me even before I'd done manual keyboard and a11y testing. 

Let's check if the focus loss issue can be replicated in trunk. 

I'll look at this tomorrow and then update tests to provide coverage.

getdave replied:
> Rather than say, "Synced with the selected %s.' it would be far more helpful to say something like "Synced with [pagename](https://github.com/WordPress/gutenberg/pull/pagelink)." Hopefully translated link interpolation here isn't too much of a pain.

@jeryj I want to note that this will be part of a follow up. Why? Because the scope of this PR is already expanding and we've already planned a "next step" design for this feature which will remove the `Link` input and replace with a UI which represents "the Page" (see PR desc for details). So in short - yes I agree with you but we should avoid that blocking this PR.

getdave replied:
> keyboard navigation is broken on the sidebar inspector for the navigation block. If you press Tab from the Text input, it moves focus to the canvas instead of the next item in the sidebar.

I tested this on `trunk` and I can see a similar issue with those fields whereby it looks like tab order is maintained when you tab out of the `Text` field. However, when you actually starting type your focus is somehow elsewhere. I suspect RichText is stealing the focus somehow or something. 

Can you replicate this?

jeryj replied:
> Let's check if the focus loss issue can be replicated in trunk.

It's a similarly broken on trunk, but not quite as severely as it is when the unsync button in the link editing field is present. On trunk, (or an unsynced navigation item), the first time tabbing through resets you to the top once you hit the link editing field. The second time it lets you tab all the way through. On this PR, you can't ever tab past the unsync edit.

My hunch is that fixing it on trunk would also fix it in this instance.

getdave committed:
Test unlinking entity

getdave committed:
Improve scope of test

getdave committed:
Improve error handling in binding

getdave committed:
Implement a11y fixes as per https://github.com/WordPress/gutenberg/pull/71630/files\#r2398999057

getdave committed:
Improve tests to provide coverage for a11y relationship between unsync button and input

getdave committed:
Extract simple component for help text and add tests

getdave committed:
LinkControl: Add accessibility association for unlink button

- Add aria-describedby association between unlink button and help text
- Generate unique helpTextId using useInstanceId hook for entity use cases
- Create custom help text with proper semantic HTML (p tag)
- Add tooltip support using showTooltip and label props
- Position help text inside search-input-wrapper with custom styling
- Use standard design system variables for consistent styling
- Update button label to 'Unsync and edit' for better UX
- Add comprehensive unit tests for accessibility association
- Fix existing test regression for button name change

Addresses a11y feedback: 'The Synced with the selected page help text
underneath this input's aria-describedby isn't attached to the button,
it's attached to an unfocusable input, so screen reader users will
likely skip it. If we're in this synced state, the aria-described by
text should get assigned to the button so it's accessible.'

getdave committed:
LinkControl: Clean up URLInput and SearchInput components

- Remove unused help prop from LinkControlSearchInput
- Add aria-describedby support to URLInput component
- Remove unused helpText variable and sprintf import
- Complete the accessibility association implementation

getdave committed:
URLInput: Remove unused ariaDescribedBy prop

- Remove ariaDescribedBy from props destructuring
- Remove ariaDescribedBy from inputProps
- This prop was added but never actually used in LinkControl
- The accessibility association is handled directly in LinkControl

getdave committed:
Implement bding in Submenu via shared code paths

getdave committed:
Register Submenu bindings

getdave committed:
Revert bindings docs

getdave committed:
Lint and fix

getdave commented on lib/compat/wordpress-6.9/block-bindings.php:
```diff
@@ -112,10 +112,12 @@ function gutenberg_block_bindings_render_block( $block_content, $block, $instanc
 function gutenberg_get_block_bindings_supported_attributes( $block_type ) {
 	// List of block attributes supported by Block Bindings in WP 6.8.
```
@ockham There's some references to 6.8. Can I get a confidence check on this file 🙏

scruffian commented on lib/compat/wordpress-6.9/entity-block-bindings.php:
```diff
@@ -0,0 +1,111 @@
+	// TODO: Add validation to ensure required attributes exist
+	// TODO: Add error handling for invalid entity types or missing data
+	// TODO: Consider property mapping between posts and terms for future keys
+	// TODO: Support additional keys beyond 'url' in the future
```
Should we remove these before merging?

getdave replied:
I thought I'd removed those. Yes! We should

scruffian reviewed (approved):
This is working really well for me. I think we should merge it and iterate in follow ups.

getdave committed:
Remove TODOs in code

getdave commented:
I agree @scruffian. I think we should merge and iterate. We know we need the work to land. So let's merge and follow up with the items noted in the Description of this PR. I can create Issues for those and tie to a main Issue if that helps.

getdave committed:
Add backport PR info

getdave committed:
Fix autoformatting breaking changelog

getdave commented on backport-changelog/6.9/10141.md:
```diff
@@ -0,0 +1,3 @@
+https://github.com/WordPress/wordpress-develop/pull/10141
```
@cbravobernal @ockham I've created this as a precaution. But I think it shouldn't land in Core as it creates too much redundancy. Let's try and work together to get the Post and Term data sources fixed to be more flexible.

t-hamano commented:
Recently, two navigation tests have become unstable.

- https://github.com/WordPress/gutenberg/issues/72065
- https://github.com/WordPress/gutenberg/issues/72064

I'm not sure if the issue is directly related to this PR, but it seems that the tests have become unstable since this PR was merged. Could someone please look into this if you have time?

cbravobernal commented:
Cherry-picked in https://github.com/WordPress/gutenberg/commit/7a93fd50caebd2437466282c0d67e2a46a27357b for Gutenberg 21.8

jeryj commented:
Flaky Test fixes are being worked on here: https://github.com/WordPress/gutenberg/pull/72132

getdave commented:
Tracking https://github.com/WordPress/gutenberg/issues/72059
